fix DistributedDataParallel for VariableImpl to TensorImpl redesign

kmatzen opened the pull request:
Summary: https://github.com/pytorch/pytorch/pull/13827 introduced a change where underlying storage could no longer be changed after calling tensor.data.  There are two places in the DistributedDataParallel class where this requirement is violated.  For now, fix up the DistributedDataParallel class by replacing the inplace storage update with a copy and removing one optimization meant to reduce the memory footprint on replicas.

Differential Revision: D13571669

soumith commented:
cc: @teng-li

teng-li commented on torch/nn/parallel/distributed.py:
```diff
@@ -391,7 +391,7 @@ def _sync_params(self):
                                          self.broadcast_bucket_size)
             for tensors, module_params_data in zip(result[1:], self.modules_params_data[1:]):
                 for tensor, param_data in zip(tensors, module_params_data):
-                    param_data.set_(tensor)
```
@kmatzen Can you elaborate on why storage could no longer be changed here?  I don't see tensor.data is called on this one

teng-li commented on torch/nn/parallel/distributed.py:
```diff
@@ -391,7 +391,7 @@ def _sync_params(self):
                                          self.broadcast_bucket_size)
             for tensors, module_params_data in zip(result[1:], self.modules_params_data[1:]):
                 for tensor, param_data in zip(tensors, module_params_data):
-                    param_data.set_(tensor)
```
Never mind, actually this is a tensor.data.  I wonder how much performance degradation this will cause, since this is on the DDP critical path.

kmatzen replied:
Is there another way with the redesign to alias tensor storage?  I didn't really understand how to do that anymore.

yf225 replied:
It would be great to measure how much performance degradation this incurs, and if the overhead is significant we can consider adding a `allow_tensor_metadata_change` flag to `.detach()`.

teng-li commented:
I wonder if you can add a simple test for this?  Like a model wrapped up in DDP (that uses two or more GPUs) that only calls the forward, and verifies that the two or more GPUs have the same parameters.

kmatzen commented:
Looks like two tests failed due to a failure to download resnet18-5c106cde.pth. Is there a way to trigger a re-run of just these two test configurations?

teng-li commented:
@pytorchbot retest this please

kmatzen committed:
fix DistributedDataParallel for VariableImpl to TensorImpl redesign (#15703)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/15703

https://github.com/pytorch/pytorch/pull/13827 introduced a change where underlying storage could no longer be changed after calling tensor.data.  There are two places in the DistributedDataParallel class where this requirement is violated.  For now, fix up the DistributedDataParallel class by replacing the inplace storage update with a copy and removing one optimization meant to reduce the memory footprint on replicas.

Existing bug manifests as the following error when each process has 2+ GPUs:
```
Traceback (most recent call last):
  File "/home/matzen/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/matzen/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "test_distributed.py", line 1438, in _run
    getattr(self, self.id().split(".")[2])()
  File "test_distributed.py", line 1377, in wrapper
    fn(self)
  File "test_distributed.py", line 95, in wrapper
    return func(*args, **kwargs)
  File "test_distributed.py", line 111, in wrapper
    return func(*args, **kwargs)
  File "test_distributed.py", line 1354, in test_DistributedDataParallel
    self._test_DistributedDataParallel(gpu_subset=gpus, rank=rank)
  File "test_distributed.py", line 1316, in _test_DistributedDataParallel
    True
  File "test_distributed.py", line 1251, in _test_DDP_5iter
    loss,
  File "test_distributed.py", line 1230, in _test_DDP_helper
    output = model(input_var)
  File "/data/users/matzen/pytorch/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/data/users/matzen/pytorch/torch/nn/parallel/distributed.py", line 362, in forward
    self._sync_params()
  File "/data/users/matzen/pytorch/torch/nn/parallel/distributed.py", line 394, in _sync_params
    param_data.set_(tensor)
RuntimeError: set_storage is not allowed on Tensor created from .data or .detach()
```

Configuring distributed tests to also run with a secondary "small world" configuration.  Test machines with at least 2 GPUs will then have more than 1 GPU assigned to a single process to exercise corner cases that are not exercised when each process only gets 1 GPU.

In order to support this testing scenario where the world size is only 1, I've also relaxed a check on public_address.  Since no other process needs to connect to the group, it's fine if the public address is left unspecified.

Differential Revision: D13571669

fbshipit-source-id: e23bff76db230d265c71d90145969cf9335c570b

pietern commented:
@kmatzen Is this still relevant?
